### ISSUE TITLE
generated packages conflict with semantic import version

### ISSUE THREAD

**Issue by carnott-snap** [golang/protobuf#945]
**What version of protobuf and what language are you using?**
Version: `v1.3.2`

**What did you do?**
I setup a library as [a gist](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369) to house the problematic behaviour:
- I created a [proto](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/6e04cf09ce9aca20b3f072082bd418c10eb85ff5#file-addressbook-proto) and [generated code that exported symbols](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/6e04cf09ce9aca20b3f072082bd418c10eb85ff5#file-addressbook-pb-go): tagged as `v1.0.0`.
- I deprecated some symbols in the [proto](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/8f96c91c1943a7b1f399aae66277fbce7d690f53#file-addressbook-proto), [`AddressBook.People`](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/8f96c91c1943a7b1f399aae66277fbce7d690f53#file-addressbook-proto-L28) and [`Person.Name`](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/8f96c91c1943a7b1f399aae66277fbce7d690f53#file-addressbook-proto-L7) and regenerated the [go code](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/8f96c91c1943a7b1f399aae66277fbce7d690f53#file-addressbook-pb-go): tagged as `v1.1.0`.
- I sunset the fields in the [proto](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/8f96c91c1943a7b1f399aae66277fbce7d690f53#file-addressbook-pb-go) and regenerated the [go code](https://gist.github.com/carnott-snap/e5680b9f9309843761fe922c6c619369/2da3324f4da82da4d8624d2eab91a4d9426ff311#file-addressbook-pb-go): tagged as `v2.0.0`.

I then tried to import both packages, [via play](https://play.golang.org/p/8BMIB4nTMvv), as promised by [semantic import versioning](https://github.com/golang/go/wiki/Modules#semantic-import-versioning):
```go
package main

import (
	"fmt"

	one "gist.github.com/e5680b9f9309843761fe922c6c619369.git"
	two "gist.github.com/e5680b9f9309843761fe922c6c619369.git/v2"
)

func main() {
	fmt.Println(one.Person{}, two.Person{})
}
```

**What did you expect to see?**
```
{ 0  [] <nil> {} [] 0} {0  [] <nil> {} [] 0}
```

**What did you see instead?**
```
panic: proto: duplicate enum registered: e5680b9f9309843761fe922c6c619369.Person_PhoneType

goroutine 1 [running]:
github.com/golang/protobuf/proto.RegisterEnum(...)
	/tmp/gopath107897050/pkg/mod/github.com/golang/protobuf@v1.3.2/proto/properties.go:458
gist.github.com/e5680b9f9309843761fe922c6c619369.git/v2.init.0()
	/tmp/gopath107897050/pkg/mod/gist.github.com/e5680b9f9309843761fe922c6c619369.git/v2@v2.0.0/addressbook.pb.go:202 +0x260
```

**Comment by carnott-snap** [comment]
Thank you for the information on the fundamental nature of the protobuf language, I am somewhat new and appreciate the context. That being said, I am still concerned about the core use case that my gist identifies: how does one handle breaking changes to generated Go packages in a language compatible manner?

Unfortunately, the reality of this situation is that we have libraries and they need to make breaking changes. Most other languages only allow one version (major, minor, patch, whatever), so this is a non-issue, but the Go language authors chose differently when they released modules. The problem is even crops up when there is no change to the proto since any breaking changes within a module will force a major version bump.

As library owners, we could ignore semver and make breaking changes whenever, or never stabilize our proto containing modules staying beneath `v1.0.0`. This has the downside that our libraries become painful to work with. Any transitive customer may get nasty conflicts when one dependency upgrades a proto containing module past a breaking change that another cannot handle. This is the reason one stabilises an API.

Alternatively, you could simply never break compatibility, as was suggested before modules were released. Unfortunately, sun setting fields is common and I have times that it needed to happen personally.

Another option is to manually ensure that if you import a given major version, you never import any other versions simultaneously. This may be feasible for a module with total control of dependencies set, but if you are relying upon any external dependencies that transitively use a proto containing module there are no guarantees that they will ever upgrade their code and unbreak your code.

My goal is not to rock the boat or push for sweeping changes. I simply have had trouble working with modules and protobufs in the past. I think it is warranted to have best practices and known pitfalls clearly documented both in this repo, say within the `README.md`, and on the protobuf site within the Go documentation, but I do not know of anything to that effect.

**Comment by dsnet** [comment]
> Most other languages only allow one version (major, minor, patch, whatever), so this is a non-issue, but the Go language authors chose differently when they released modules. The problem is even crops up when there is no change to the proto since any breaking changes within a module will force a major version bump.

I'm not sure what you mean by this. Go allows different *major* versions to be linked into a binary, but only a single *minor* and *patch* version for the same given *major* version. This is generally how things are in other ecosystems.

Backing up a bit. I think the issue with semver is a red herring. The issue at hand seems to be that the same `.proto` file or a clone variant of the same `.proto` file is being used to generate into at least two different Go package. Such a use case represents a clash in the protobuf namespace.

Have you tried using a different proto package namespace for the `.proto` files associated with v2? If not, why does that not fix the issue for you?

**Comment by carnott-snap** [comment]
> I'm not sure what you mean by this. Go allows different _major_ versions to be linked into a binary, but only a single _minor_ and _patch_ version for the same given _major_ version. This is generally how things are in other ecosystems.
Apologies for not being more concrete, I was thinking of how in Java you cannot link `v1.2.3` and `v2.3.4` in the same runtime, lest you get class path import issues. However on further review, it appears this differs widely, since JavaScript, specifically npm, will let you link even _minor_ and _patch_ versions in parallel.

> Backing up a bit. I think the issue with semver is a red herring. The issue at hand seems to be that the same `.proto` file or a clone variant of the same `.proto` file is being used to generate into at least two different Go package. Such a use case represents a clash in the protobuf namespace.
I understand that from the perspective of the Go toolchain v0/v1 and v2 are separate packages, but this is not how I, or others I have worked with, intuitively understand versions: v2 is just a breaking API change to v1. This distinction is confusing and I think warrants being called out in the official documentation. It would be nice if the generated Go code maintained the wire compatibility versioning guarantees, but this may be impossible or a breaking change to this library.
 
> Have you tried using a different proto package namespace for the `.proto` files associated with v2? If not, why does that not fix the issue for you?
We use protobuf because of the ability to make (wire format) compatible sunset and deprecation changes that will not impact our customers. It confuses me that you are suggesting we ignore this protobuf feature, place more friction on our customers, and create new APIs for every breaking change. Do I misunderstand? If not, can this suggestion be documented because I have seen no warnings or caution around `reserved` fields. Furthermore, if this feature is in fact deprecated, as I think you imply, would it make sense to track its sunset for `proto4`?

Unfortunately, our use case is a little more nuanced. Some of the protos that we consume are shared resources, and we cannot control namespacing or prevent backwards incompatible changes, without breaking the toolchains and libraries we depend on. If my inferences above are correct, and this is an official stance, it would be nice to have this documented, so that can work with those maintainers to resolve this problem upstream.

**Comment by carnott-snap** [comment]
@dsnet, I am sorry for how long this discussion took, but thoroughly appreciate your patience and detail. Unfortunately, the differences between the current state of the proto3 spec, especially as they are being used by our upstream dependencies, and the suggestions you made are sizeable. We would like to work with our upstream peers and get your suggested best practices enacted, but I am concerned without something more formal, we will not have the leverage to enact the changes you describe.
